api/src/main/java/net/kyori/adventure/text/Component.java

```diff
@@ -146,17 +146,57 @@
    */
   @Contract(value = "_, _ -> new", pure = true)
   static @NonNull TextComponent join(final @NonNull ComponentLike separator, final Iterable<? extends ComponentLike> components) {
+    return joinWithFinalSeparator(separator, separator, components);
+  }
+
+  /**
+   * Joins {@code components} using {@code separator} with {@code finalSeparator} between the last two components.
+   *
+   * @param separator the separator
+   * @param finalSeparator the final separator
+   * @param components the components
+   * @return a text component
+   * @since 4.8.0
+   */
+  @Contract(value = "_, _, _ -> new", pure = true)
+  static @NonNull TextComponent joinWithFinalSeparator(final @NonNull ComponentLike separator, final @NonNull ComponentLike finalSeparator, final @NonNull ComponentLike@NonNull... components) {
+    return joinWithFinalSeparator(separator, finalSeparator, Arrays.asList(components));
+  }
+
+  /**
+   * Joins {@code components} using {@code separator} with {@code finalSeparator} between the last two components.
+   *
+   * @param separator the separator
+   * @param finalSeparator the final separator
+   * @param components the components
+   * @return a text component
+   * @since 4.8.0
+   */
+  @Contract(value = "_, _, _ -> new", pure = true)
+  static @NonNull TextComponent joinWithFinalSeparator(final @NonNull ComponentLike separator, final @NonNull ComponentLike finalSeparator, final @NonNull Iterable<? extends ComponentLike> components) {
     final Iterator<? extends ComponentLike> it = components.iterator();
     if(!it.hasNext()) {
       return Component.empty();
     }
+
     final TextComponent.Builder builder = text();
-    while(it.hasNext()) {
-      builder.append(it.next());
-      if(it.hasNext()) {
-        builder.append(separator);
+    ComponentLike component = it.next();
+    while(true) {
```

Comment:
Is there any other way to do this? This feels a bit ugly.

Reply:
I can't *think* of any way to do it without a sized collection that we can iterate over with an indexed loop. I suppose we could make this take a list instead of an iterable? There might be another way to write it, perhaps iterating while the component variable isn't null and setting it to null when the iterator runs out of elements? Or perhaps just replacing this with `it.hasNext()`, but then there's duplicate calls to it which isn't a massive issue I suppose.